### PR TITLE
Injection Inception, Change #6: Final cleanup of test environment setup

### PR DESCRIPTION
**Pull request by diosmosis**
This PR is based off of https://github.com/piwik/piwik/pull/8008

Changes:
- [X] Rename TestingEnvironment to TestingEnvironmentVariables, since that is really all it is for.
- [X] Remove TestingEnvironment::addHooks() and move to new environment manipulator (along w/ Environment.bootstrapped event observer in test.php).
- [x] Get rid of the TestingEnvironment test event TestingEnvironment.addHooks and replace w/ DI use.
- [x] Merge MockAccess w/ FakeAccess.
- [x] Resolve redundant Fixture/TestCase variable `$loadTranslations` (handled byTestingEnvironmentVariables.
- [x] Add custom definition source TestingEnvironmentVariablesDefinitionSource that allows accessing TestingEnvironmentVariables variables through DI. Also allows overriding these variables easily from within test case classes (via provideContainerConfig(BeforeClass)).

### CODE REVIEW

**Comment by diosmosis** [comment]
After merging, QueuedTracking will have to be modified, as it still uses the TestingEnvironment.addHooks event.

**Comment by diosmosis** [comment]
Looks like this PR depends on #8026, there are some odd access related failures. Waiting for that one.

File: core/Application/Environment.php [hunk 1]
```diff
@@ -185,4 +191,27 @@ private function getGlobalSettingsProviderOverride()
...
     }
+
+    private function invokeBeforeContainerCreatedHook()
+    {
+        foreach (self::$globalEnvironmentManipulators as $manipulator) {
+            $manipulator->beforeContainerCreated();
+        }
+    }
+
+    private function getExtraDefinitionsFromManpulators()
```
Comment (mnapoli) [inline core/Application/Environment.php]: typo: `getExtraDefinitionsFromManpulators` -> `getExtraDefinitionsFromManipulators`

File: misc/cron/updatetoken.php [hunk 2]
```diff
@@ -35,7 +35,7 @@
 if ($testmode) {
     define('PIWIK_TEST_MODE', true);
 
-    Tests\Framework\TestingEnvironment::addHooks();
+    Tests\Framework\TestingEnvironmentVariables::addHooks();
```
Comment (mnapoli) [inline misc/cron/updatetoken.php]: I thought that `addHooks()` was removed? From the PR description:

> Get rid of the TestingEnvironment test event TestingEnvironment.addHooks and replace w/ DI use.

File: tests/PHPUnit/Framework/Fixture.php [hunk 3]
```diff
@@ -388,6 +387,8 @@ public static function installAndActivatePlugins()
                 $pluginsManager->activatePlugin($name);
             }
         }
+
+        $pluginsManager->loadPluginTranslations();
```
Comment (mnapoli) [inline tests/PHPUnit/Framework/Fixture.php]: Previously it was calling `Piwik\Translate::loadAllTranslations()`, which was doing:

```php
Translate::loadCoreTranslation();
$pluginsManager->loadPluginTranslations();
```

You replaced it with `$pluginsManager->loadPluginTranslations()` only.

Could it be the source of the problem you had with translations? (i.e. core translations not loaded?)

**Comment by mnapoli** [comment]
Coming back again on the topic of "environment manipulators": the [interface comment says](https://github.com/piwik/piwik/blob/test_env_di_8/core/Application/EnvironmentManipulator.php#L15-L16):

> Only used by the testing environment setup code, shouldn't be used anywhere else.

So if I understand correctly:

- we have an interface that is meant to be implemented only by one specific class
- the `Environment` has an array of "manipulators" ([link](https://github.com/piwik/piwik/blob/test_env_di_8/core/Application/Environment.php#L46)) whereas it should contain only one entry

If that's correct then maybe `Environment` could have a single object instead of an array, and we could get rid of the interface?

---

Regarding `TestingEnvironmentVariablesDefinitionSource`, it's a nice and clean implementation I'm happy to see that extending PHP-DI works so easily ;) However would it make more sense (i.e. simpler) to just take all the testing env variables, dump them into an array (and we add the prefix here) and just pass the array to be added to the container by the environment?

That would avoid the custom definition source + [this custom code for tests in the ContainerFactory]( https://github.com/piwik/piwik/blob/test_env_di_8/core/Container/ContainerFactory.php#L96-101)

File: misc/cron/updatetoken.php (hunk 2 above)
Comment (diosmosis) [inline misc/cron/updatetoken.php]: Missed this file, there are too many entry points :)

File: tests/PHPUnit/Framework/Fixture.php (hunk 3 above)
Comment (diosmosis) [inline tests/PHPUnit/Framework/Fixture.php]: I thought this happened automatically, but I can check again.

The problem I had w/ translations is that some tests require translations (like many system tests), and some tests don't test against actual translations. Disabling it or enabling it for everyone breaks some tests. I think I will go towards testing against tokens by default.

**Comment by diosmosis** [comment]
> If that's correct then maybe Environment could have a single object instead of an array, and we could get rid of the interface?

We could just have one, I was hoping an array would look like less of a hack.

As for the interface, I think it's a good idea to keep it. Otherwise, Environment.php would reference a test file, namely the environment manipulator I added. I don't think application code should do that if possible.

> However would it make more sense (i.e. simpler) to just take all the testing env variables, dump them into an array (and we add the prefix here) and just pass the array to be added to the container by the environment?

Where would this be done? Would we be able to override the vars in `provideContainerConfig()`?

Note: It would only avoid the lines that add the definition source.

**Comment by mnapoli** [comment]
> Where would this be done? Would we be able to override the vars in provideContainerConfig()?

I'm not sure. But since the testing env manipulator does provide container entries to the environment (e.g. the entries defined in fixtures or test classes), it could also provide those entries through an array instead of a custom definition source? Maybe it's not possible, I was deep into the code yesterday but now I'm out of it and maybe it doesn't happen at the same time…

For the interface I'm OK to keep it I see your point with coupling the core code to test code, for the array of env manipulators I would say let's turn it into a single dependency it's that's ok with you? I'd rather keep this core code as simple as we can so that it's easy to understand

**Comment by diosmosis** [comment]
>  Maybe it's not possible, I was deep into the code yesterday but now I'm out of it and maybe it doesn't happen at the same time…

It's not easily possible it seems. Using a definition source, we can return null for variables that don't exist in the json file. Can't do that if we just add the variables in an array.

**Comment by diosmosis** [comment]
Ready for another review.

**Comment by mnapoli** [comment]
Looks good to merge, except for the the Travis build: 2 failures in SystemTests: https://travis-ci.org/piwik/piwik/jobs/65642678#L524 Is it the random failure or something else?

**Comment by diosmosis** [comment]
Ah those are queued tracking errors, I have to modify the plugin...

**Comment by diosmosis** [comment]
QueuedTracking PR: https://github.com/piwik/plugin-QueuedTracking/pull/8

The build won't pass, there's one failure left. I tried to fix it, but can't seem to figure out what the problem is.

**Pull request merged**